### PR TITLE
[MM-51054] Allow enabling/disabling plugins through env override

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR implements some logic to enable overriding the plugins state through env variables more granularly (at plugin level). Before it was only possible to set the whole JSON object (e.g. `MM_PLUGINSETTINGS_PLUGINSTATES="{\"com.mattermost.calls\":{\"Enable\":true}"`) which isn't very convenient.

To be noted, this won't behave exactly like other environment overrides as the implementation is lacking a way to detect the applied override and remove it from config so it's more of a permanent override meaning the value gets actually persisted in the config. I think this is fine for the current need (permanently disable certain plugins) and can be improved at a later time.

@amyblais This is needed to unblock https://community.mattermost.com/private-core/channels/edwwn1bfafgzxxsufk8znst7ye

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-51054

#### Release Notes

```release-note
Added support to override plugins state (enable/disable) through environment variables (e.g. `MM_PLUGINSETTINGS_PLUGINSTATES_PLAYBOOKS=true`). Note: the state will be persisted in the configuration so it should not be used for temporary overrides.
```

### CODE REVIEW

**Review by cpoile — APPROVED** [review]
Nice

**Review by isacikgoz — APPROVED** [review]
LGTM, thanks for adding the test!

**Comment by lieut-data** [comment]
Awesome work, @streamer45! One thought: is there a reason we don't hard-code this restriction directly into the build for `focalboard`? I'm considering doing this for playbooks in the upcoming monorepo transition, and it has the nice advantage of a) being impossible to misconfigure, and b) allowing peer nodes that haven't upgraded to keep running playbooks (& serving requests!)

**Comment by streamer45** [comment]
> Awesome work, @streamer45! One thought: is there a reason we don't hard-code this restriction directly into the server build for `focalboard`? I'm considering doing this for playbooks in the upcoming monorepo transition, and it has the nice advantage of a) being impossible to misconfigure, and b) allowing peer nodes that haven't upgraded to keep running playbooks (& serving requests!)

@lieut-data 

Honestly I am not sure, as depending on where you were thinking of adding this hard-coded restriction it may or may not work as we expect in this instance (yet to be tested of course). The issue we are targeting with this PR requires the plugin to be disabled in config because as soon as the server starts the various services it may be too late. In fact there's already some logic to do this but it's failing possibly due to a race condition in cluster nodes interaction.

But overall I agree that there's likely a simpler way to disable specific plugins that are moving to being product but I'd defer to their owners to come up with a proper solution. I think these changes are almost orthogonal to that effort which should definitely continue.

Also, it looks like there's some other work in the making to disable more boards plugin related logic (https://github.com/mattermost/mattermost-server/pull/22377) so it makes sense to coordinate in such a way that we don't do extra work and keep consistency when porting over other products.

**Comment by streamer45** [comment]
@amyblais To which branch should we cherry-pick this to get it tested on Cloud for release?

**Comment by amyblais** [comment]
``cloud`` branch, and I'm not sure if this also needed for ``release-7.9`` branch.

**Comment by streamer45** [comment]
> `cloud` branch, and I'm not sure if this also needed for `release-7.9` branch.

Not strictly needed but may as well do both since we are here.

**Comment by lieut-data** [comment]
> But overall I agree that there's likely a simpler way to disable specific plugins that are moving to being product but I'd defer to their owners to come up with a proper solution.

I think of this not as "disabling a plugin", which has all sorts of secondary effects (e.g. cluster awareness), but instead "refusing to run a plugin". Since enabling product mode, the server should always refuse to run the focalboard plugin, regardless of configuration. (This seems to be what keeps tripping us up, and while we now have a new way to disable it... we're reliant on this being set correctly.)

Another way to tackle this might be to skip MPA (again) for v7.9, and only ship the monorepo. It seems weird to have to mentally support a single release that ships boards different from all past and all future releases...

cc @streamer45, @isacikgoz, @wiggin77

**Comment by streamer45** [comment]
Absolutely, I am just trying to be cautious here because disabling a plugin is one thing, changing/overriding the existing logic is another and since we couldn't really reproduce the issue locally (not yet anyway) I am not confident enough to make some changes that may or may not work. 

So I think of this PR as mostly unrelated to products but merely focusing on unblocking the upcoming Cloud release. I think what you suggests is totally valid and likely the ideal solution but I also feel it should be brought forward with a little more thought and coordination, not in the rush of an incident so to speak. Does that make sense?

**Comment by lieut-data** [comment]
> ... but I also feel it should be brought forward with a little more thought and coordination, not in the rush of an incident so to speak. Does that make sense?

@streamer45, in the isolation of this incident, I totally agree. But at the same time, it feels like we're on attempt #4 or #5 to "disable boards plugin via configuration" instead of just hard-coding the outcome we expect. I'd love to build confidence with Cloud, hence the different tack being proposed.

I can help continue the conversation with @agnivade on Monday as well, and we can build on this change one way or another.

**Comment by wiggin77** [comment]
I agree with @lieut-data in that we should simply ensure Boards plugin does not run during server start-up.  It should not require env vars or any other user set up.  Currently the plugin is disabled only after allowing it to be loaded, which is the cause of the deployment problems in cloud.  It should simply not be allowed to run, and should not call `disablePlugin` which triggers the websockets events that cause cluster nodes to error.

We originally talked about having a general plugin black-list that would be useful ongoing to kill off plugins.  I still think that would have been the best approach.

**Comment by agnivade** [comment]
It is not that simple. The problem is in the upgrade procedure. We cannot just skip running boards, we need to disable it. Otherwise, it will appear confusing and incorrect in the system console. And the moment we disable it, the config disable event goes across the cluster. And in a rolling upgrade scenario, it goes to the other node which doesn't have product mode related code yet.

I think the current approach is okay.

**Comment by isacikgoz** [comment]
@lieut-data @agnivade I think we should verify this solution and unblock the release, WDYT?

**Comment by agnivade** [comment]
Further discussion happened in the channel.

**Pull request merged**
